src/dune_cache_storage/layout.ml

```diff
@@ -82,7 +82,12 @@ let value_storage_dir = Versioned.value_storage_dir Version.Value.current
 let value_path = Versioned.value_path Version.Value.current
 
 let create_cache_directories () =
-  List.iter
-    [ temp_dir; metadata_storage_dir; file_storage_dir; value_storage_dir ]
-    ~f:(fun path -> ignore (Fpath.mkdir_p (Path.to_string path) : Fpath.mkdir_p_result))
+  [ temp_dir; metadata_storage_dir; file_storage_dir; value_storage_dir ]
+  |> List.fold_left ~init:(Ok ()) ~f:(fun result path ->
+    match result with
+    | Error _ as e -> e
```

Comment:
A little suspect to use a fold here but to terminate on the very first failure

Reply:
I want to only continue if all of them have been successfully created. The other option I could think of would be to `map` and then something like `Result.List.all_unit` but there is no point in continuing to create folders if one of them has failed. Likewise, I also don't think there is a point to delete them again if all but the last have succeeded.

(Looking at `Stdune.Result` we have `Result.List.fold_left` which I guess inlines the match on the result, so that can be made a bit nicer but it does essentially still the same).

Reply:
I think you want `Result.List.iter` here. If we don't have that, you can add it.

I'm not sure what you mean by inlining, but my point is that using `List.fold_left` will always traverse the entire list even though you trying to step after the first failure.

```ocaml
let x = [ false ; true ]
let f x = if x then Error () else Ok ()
let () = List.fold_left x ~init:(Ok ()) ~f:(fun x -> print_endline "foo"; f x)
let () = Result.List.fold_left x ~init:() ~f:(fun x -> print_endline "foo"; f x)
```

The `List.fold_left` prints `foo` twice, while `Result.List.fold_left` prints `foo` only once. I think you intend to have the latter behavior.

Reply:
Sure, but that's why the first statement in the body is a `match` so that a previous `Error` will short-circuit into returning `Error` and not call any more `mkdir`.

Ah yes, `Result.List.iter` is exactly what I wanted, thus I can avoid match in `~f`.